src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenNullCheckedParameterTests.cs

```diff
@@ -1321,6 +1321,44 @@ .maxstack  3
 }");
         }
 
+        [Fact]
+        public void TestNullCheckedAsyncIterator()
+        {
+            var source = @"
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+class C
+{
+    async IAsyncEnumerable<char> GetChars(string s!!, Task t)
+    {
+        foreach (var c in s)
+        {
+            await t;
+            yield return c;
+        }
+    }
+}";
+            var compilation = CompileAndVerify(CreateCompilationWithTasksExtensions(new[] { source, AsyncStreamsTypes }, parseOptions: TestOptions.RegularPreview));
+            compilation.VerifyIL("C.GetChars", @"
+{
+  // Code size       33 (0x21)
+  .maxstack  3
+  IL_0000:  ldc.i4.s   -2
+  IL_0002:  newobj     ""C.<GetChars>d__0..ctor(int)""
+  IL_0007:  ldarg.1
+  IL_0008:  ldstr      ""s""
+  IL_000d:  call       ""ThrowIfNull""
```

Comment:
I was a bit surprised to see us create the state machine then do the `null` check. Given that we do the `null` check before calling `base / this` ctor would've expected the `null` check first here. 

Not blocking if everyone else feels that way, can deal with it in a follow up.

Reply:
A few test plan items are marked "address after merge". I think it would be fine to add a similar item to the test plan #36024 to see if we can improve this.